### PR TITLE
Adding an SortedArray data structure

### PR DESCRIPTION
**Pull request by TheStefan**
Implements all meaningful operations of an array list for a sorted array list. Contains test code, documentation and source.

### CODE REVIEW

**Comment by fragglet** [comment]
Hi! Thanks for submitting this. I have a few comments.

File: src/sortedarray.h [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+ * @ref sortedarray_append, or @ref sortedarray_insert.
+ *
+ * To remove a value from a SortedArray, use @ref sortedarray_remove
+ * or @ref sortedarray_remove_range.
+ */
+
+#ifndef ALGORITHM_SORTEDARRAY_H
+#define ALGORITHM_SORTEDARRAY_H
+
+#include "arraylist.h"
```
Comment (fragglet) [inline src/sortedarray.h]: A fundamental design decision of this library is that each pair of .c/.h files should be standalone. So it should be possible to take one of these files and just add it into a project without needing to depend on the entire library. Please refactor the code so that this is usable standalone without depending on arraylist.

File: test/test-sortedarray.c [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+    }
+
+    return (*f1 == *f2) ? 1 : 0;
+}
+
+void check_sorted_prop(SortedArray *sortedarray)
+{
+    unsigned int i;
+    for (i = 1; i < sortedarray->array->length; i++) {
+        float *pf = *(sortedarray->array->data + i);
```
Comment (fragglet) [inline test/test-sortedarray.c]: Please just use array syntax, ie.

    sortedarray->array->data[i]

File: test/test-sortedarray.c [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+
+    if (f1 == f2) {
+        return 1;
+    }
+
+    if (f1 == NULL || f2 == NULL ) {
+        return 0;
+    }
+
+    return (*f1 == *f2) ? 1 : 0;
```
Comment (fragglet) [inline test/test-sortedarray.c]: Just

    return *f1 == *f2

File: test/test-sortedarray.c [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+#include "alloc-testing.h"
+#include "framework.h"
+
+#include "sortedarray.h"
+
+#define TEST_SIZE 100
+
+int compare_float(void* v1, void* v2) 
+{
+    float* f1 = (float*) v1;
```
Comment (fragglet) [inline test/test-sortedarray.c]: Please put the * next to the variable name, not the type name. If you could check through all your code and fix anywhere else you've done this that would be good.

File: test/test-sortedarray.c [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+        float *pf = *(sortedarray->array->data + i);
+        free(pf);
+    }
+
+    sortedarray_free(sortedarray);
+}
+
+SortedArray *generate_sortedarray(void)
+{
+    //generate a sorted array of length TEST_SIZE, filled with random numbers
```
Comment (fragglet) [inline test/test-sortedarray.c]: Please change your code to use /* ... */ comments everywhere, not // comments, for consistency with the rest of the library.

File: test/test-sortedarray.c [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+#include <time.h>
+
+#include "alloc-testing.h"
+#include "framework.h"
+
+#include "sortedarray.h"
+
+#define TEST_SIZE 100
+
+int compare_float(void* v1, void* v2)
```
Comment (fragglet) [inline test/test-sortedarray.c]: I'm not sure why you structured your tests around float arrays. The library already has functions for testing equality/comparison of ints. Floats are rather problematic because they cannot be reliably compared for equality. It would be better if you just restructured your tests to work with floats.

File: src/sortedarray.c [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+                              unsigned int length)
+{
+    //just remove, because sorted prop will not be violated
+    arraylist_remove_range(sortedarray->array, index, length);
+}
+
+int sortedarray_insert(SortedArray *sortedarray, SortedArrayValue data)
+{
+    //do a binary search like loop to find right position
+    unsigned int b = 0;                             //left search index
```
Comment (fragglet) [inline src/sortedarray.c]: These short variable names make the code hard to read, and I don't quite understand why you used 'b', 'e' and 'r'. Probably 'left' and 'right' would be more readable.

File: src/sortedarray.h [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+
+/**
+ * A value to store in @ref SortedArray.
+ */
+typedef ArrayListValue SortedArrayValue;
+
+/**
+ * A SortedArray structure. Use @ref sortedarray_new to create one.
+ *
+ * The SortedArray is basically an ArrayList with data in sorted order. User-
```
Comment (fragglet) [inline src/sortedarray.h]: "basically an ArrayList" refers to the other file in the library, and the documentation should be able to stand on its own as well as the code. A better description might be something like:

"A SortedArray is an array of elements which are always maintained in a consistent order."

Might also be worth including estimates for the time needed for lookups/insertions (O(log n) I think).

File: src/sortedarray.h [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+int sortedarray_insert(SortedArray *sortedarray, SortedArrayValue data);
+
+/**
+ * Find the index of a value in a SortedArray.
+ *
+ * @param sortedarray   The SortedArray to find in.
+ * @param data          The value to find.
+ * @return              The index of the value or -1 if the value is not found.
+ */
+int sortedarray_index_of(SortedArray *sortedarray, SortedArrayValue data);
```
Comment (fragglet) [inline src/sortedarray.h]: There's index_of but no apparent API to retrieve a particular element by its index? That seems like a fundamental operation.

File: src/sortedarray.c [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+ * expected results.
+ *
+ * @param sortedarray   The SortedArray.
+ * @param data          The value to compare to.
+ * @param b             The start of the range (inclusive).
+ * @param e             The end of the range (inclusive).
+ *
+ * @return The first index where cmp_func returns 0.
+ */
+static unsigned int sortedarray_first_index(SortedArray *sortedarray,
```
Comment (fragglet) [inline src/sortedarray.c]: Avoid function prototypes where possible: move these functions up to the top of the file. It's not necessary to attach full doc comments for internal (static) functions.

**Comment by fragglet** [comment]
Additional comment: the indentation style of your code is not consistent with the rest of the library. I use tabs for indentation, spaces for alignment. See here for details:

https://blog.codinghorror.com/death-to-the-space-infidels/

**Comment by TheStefan** [comment]
Okay, I'll fix the comments.

About the indentation: I suppose you mean I should use the third option from the blog article.

**Comment by fragglet** [comment]
Exactly - thanks!

File: test/test-sortedarray.c (hunk 6 above)
Comment (TheStefan) [inline test/test-sortedarray.c]: I understand what you mean. The sortedarray structure supports separate comparison and equality functions. So I use floats to make the non fractional part used for sorting and the fractional part for equality. My index of is structured in such way that it needs to find the equal object, not the object which compares to be equal. Since I do not do any arithmetic with floats comparisons won't be a problem, however it is indeed wiser to use ints since it is less likely to introduce a bug. However to keep the special case for elements which compare equal which are not equal I would not be able to use the already built in functions.

**Comment by TheStefan** [comment]
I have changed all the things you have commented. If there are still some problems, please let me know!

**Comment by sgtbigman** [comment]
This is neat. I will definitely be looking at the code to learn some things! 

Just as a note: I'm not sure if the maintainer of this project will care or not, but I wanted to let you know that you may want to do a git rebase before he merges this in to master. In doing so, you can squash/fixup some of your commits into other ones; for example, you could make it so that your code had the correct indentation style from the beginning, instead of being fixed in a later commit.

That note aside, again, neat work. :)

**Comment by TheStefan** [comment]
Thank you! I'll look into rebasing

**Comment by fragglet** [comment]
Sorry I still haven't got back to you. I'll see if I can take a look today.

File: src/sortedarray.c [hunk 11]
```diff
@@ -0,0 +1,273 @@
...
+
+	while (left < right) {
+		index = (left + right) / 2;
+
+		int order = sortedarray->cmp_func(data, 
+		                                  sortedarray->data[index]);
+		if (order > 0) {
+			left = index + 1;
+		}
+		else {
```
Comment (fragglet) [inline src/sortedarray.c]: Should be on same line, ie.

    } else {

File: src/sortedarray.c [hunk 12]
```diff
@@ -0,0 +1,273 @@
...
+}
+
+int sortedarray_insert(SortedArray *sortedarray, SortedArrayValue data)
+{
+	/* do a binary search like loop to find right position */
+	unsigned int left  = 0;
+	unsigned int right = sortedarray->length;
+	unsigned int index = 0;
+
+	right = (right > 1) ? right : 0;
```
Comment (fragglet) [inline src/sortedarray.c]: Why is this line needed? I don't understand why you need special casing for lists of length 1.

File: src/sortedarray.c [hunk 13]
```diff
@@ -0,0 +1,273 @@
...
+			left = index + 1;
+		}
+		else {
+			/* value should be at index */
+			left = right = index;
+			break;
+		}
+	}
+
+	if (sortedarray->length > 0 && sortedarray->cmp_func(data, sortedarray->data[index]) > 0) {
```
Comment (fragglet) [inline src/sortedarray.c]: This block looks strange. Do you need it?

File: src/sortedarray.c [hunk 14]
```diff
@@ -0,0 +1,273 @@
...
+		index++;
+	}
+
+	/* insert element at r */
```
Comment (fragglet) [inline src/sortedarray.c]: Comment ("at r") is wrong. Can you factor out this block into a separate function, eg.  "sortedarray_grow_to_minimum_size"?

File: src/sortedarray.c [hunk 15]
```diff
@@ -0,0 +1,273 @@
...
+	}
+
+	/* move all other elements */
+	memmove(&sortedarray->data[index + 1],
+	        &sortedarray->data[index],
+	        (sortedarray->length - index) * sizeof(SortedArrayValue));
+
+	/* insert entry */
+	sortedarray->data[index] = data;
+	++(sortedarray->length);
```
Comment (fragglet) [inline src/sortedarray.c]: There's no need for parens here.

File: test/test-sortedarray.c [hunk 16]
```diff
@@ -0,0 +1,225 @@
...
+#define TEST_REMOVE_EL 15
+#define TEST_REMOVE_RANGE 7
+#define TEST_REMOVE_RANGE_LENGTH 4
+
+void check_sorted_prop(SortedArray *sortedarray)
+{
+	unsigned int i;
+	for (i = 1; i < sortedarray->length; i++) {
+		assert(int_compare(
+					sortedarray->data[i - 1],
```
Comment (fragglet) [inline test/test-sortedarray.c]: Formatting is super weird here.

File: test/test-sortedarray.c [hunk 17]
```diff
@@ -0,0 +1,225 @@
...
+	/* assert */
+	for (i = 0; i < TEST_REMOVE_RANGE_LENGTH; i++) {
+		assert(*((int*) sortedarray->data[TEST_REMOVE_RANGE + i]) == new[i]);
+	}
+
+	check_sorted_prop(sortedarray);
+	free_sorted_ints(sortedarray);
+}
+
+void test_sortedarray_index_of(void) {
```
Comment (fragglet) [inline test/test-sortedarray.c]: { for functions should be on the following line.

File: test/test-sortedarray.c [hunk 18]
```diff
@@ -0,0 +1,225 @@
...
+#include <time.h>
+
+#include "alloc-testing.h"
+#include "framework.h"
+
+#include "compare-int.h"
+#include "sortedarray.h"
+
+#define TEST_SIZE 20
+#define TEST_ARRAY {10, 12, 12, 1, 2, 3, 6, 7, 2, 23, 13, 23, 23, 34, 31, 9,\
```
Comment (fragglet) [inline test/test-sortedarray.c]: This is only used once in the file. Just inline it.

File: src/sortedarray.h [hunk 19]
```diff
@@ -0,0 +1,210 @@
...
+ * 			value2, zero if they compare equal, or greater than
+ * 			zero if value1 compares greate than value2.
+ */
+typedef int (*SortedArrayCompareFunc)(SortedArrayValue value1,
+                                      SortedArrayValue value2);
+
+/**
+ * Definition of a @ref SortedArray
+ */
+struct _SortedArray {
```
Comment (fragglet) [inline src/sortedarray.h]: In general I prefer to make the structs opaque, ie. declare them inside the .c file and don't expose them publicly. Can you do this please?

You'll need to turn SORTEDARRAY_GET into a proper accessor function, but that's okay - the result should be more readable anyway.

**Comment by fragglet** [comment]
Thanks for reworking this - it looks much better!

I noticed that the coverage analysis for the new file could be improved:

    Lines executed:89.91% of 109

Any chance we can get closer to 100%?

**Comment by fragglet** [comment]
Ping.

**Comment by TheStefan** [comment]
I did not forget. I just currently have a lot things todo.

**Comment by fragglet** [comment]
Cool, thanks :) No pressure.

**Comment by TheStefan** [comment]
I'll take a look at it

**Comment by fragglet** [comment]
Hey, I wasn't aware you had pushed new commits since you didn't comment. Feel free to reopen if you'd still like to merge this.

**Comment by TheStefan** [comment]
Lol, next time I will give a ping.

**Comment by fragglet** [comment]
Thanks!

**Comment by TheStefan** [comment]
So I suppose that I can close it now?

**Comment by TheStefan** [comment]
:)
